### PR TITLE
Fix Python workflow for WordPieceTokenizer

### LINKED ISSUE #2948: [PyGrain Migration] 4. Migrate WordPieceTokenizer - tokenize/detokenize path
In keras_hub/src/tokenizers/word_piece_tokenizer.py, implement _tokenize_python
Implement _detokenize_python
Constructor: don't require tf-text (_allow_python_workflow=True default, match byte_pair_tokenizer.py).

### PR DESCRIPTION
**Pull request by maitry63**
This PR adds a Python workflow for `WordPieceTokenizer`, allowing tokenization and detokenization to run without requiring `tensorflow_text`.

Fixes: #2948

### Changes :

- Added pure-Python pre-tokenization.
- Added a pure-Python greedy longest-match-first WordPiece implementation.
- Added pure-Python detokenization.
- Added Python-side vocabulary and token-to-ID lookups that do not depend on
  `tensorflow_text`.
- Made `tensorflow_text` optional for the Python workflow.
- Preserved the existing `tensorflow_text` workflow.
- Made the Python workflow the default when not inside a `tf.function`.
- Added separate Python input canonicalization for tokenization and
  detokenization.
- Handled scalar TensorFlow string tensors correctly when converting their
  `.numpy()` values from `bytes` to Python `str`.

### Tests :

- Default Python workflow.
- Python workflow without `tensorflow_text`.
- Explicit TensorFlow workflow.
- Python and TensorFlow workflow equivalence.
- Python detokenization.
- Special tokens.
- Lowercasing and accent stripping.
- CJK tokenization.
- Integer and string output dtypes.
- Scalar and batched inputs.

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/tokenizers/word_piece_tokenizer.py [hunk 1]
```diff
@@ -190,6 +228,141 @@ def pretokenize(
...
+                [special_tokens_pattern, keep_split_pattern]
+            )
+        # A single capturing group around `keep_split_pattern` means
+        # `regex.split` keeps those matches as list items, while matches of
+        # the non-capturing whitespace-only alternative are dropped, just
+        # like `tf_text.regex_split(..., keep_delim_regex_pattern=...)`.
+        split_regex = re.compile(
+            f"({keep_split_pattern})|(?:{WHITESPACE_REGEX})"
+        )
+        words = [piece for piece in split_regex.split(text) if piece]
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer.py:311]: Unicode whitespace — Python's \s is Unicode-wide; RE2's is ASCII-only. "the\u00a0fox" → Python [the, fox], TF [UNK] (same for U+2009, U+3000). Fix: a Python-side whitespace pattern using the regex module's scoped ASCII flag, (?a:\s)|\p{Cc}|\p{Cf}.

File: keras_hub/src/tokenizers/word_piece_tokenizer.py [hunk 2]
```diff
@@ -190,6 +228,141 @@ def pretokenize(
...
+    if lowercase:
+        if special_tokens_pattern is not None:
+            # Do not lowercase special tokens, e.g. `"[CLS]"`.
+            special_tokens_regex = re.compile(f"^(?:{special_tokens_pattern})$")
+            words = [
+                word if special_tokens_regex.match(word) else word.casefold()
+                for word in words
+            ]
+        else:
+            words = [word.casefold() for word in words]
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer.py]: str.casefold() ≠ tf_text.case_fold_utf8 (NFKC_CaseFold). "ＦＯＸ fox" → Python [UNK, fox], TF [fox, fox]. Fix: unicodedata.normalize("NFKC", word.casefold())

File: keras_hub/src/tokenizers/word_piece_tokenizer.py [hunk 3]
```diff
@@ -190,6 +228,141 @@ def pretokenize(
...
+
+    return words
+
+
+def word_piece_tokenize_word_python(
+    word,
+    vocabulary_set,
+    unknown_token,
+    suffix_indicator,
+    max_input_chars_per_word=200,
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer.py]: Python bails at 200 chars; FastWordpieceTokenizer defaults to max_bytes_per_word=100. A 120-byte word tokenizes on Python, [UNK] on TF. Fix: len(word.encode("utf-8")) > 100

File: keras_hub/src/tokenizers/word_piece_tokenizer_test.py [hunk 4]
```diff
@@ -250,6 +251,200 @@ def test_no_splitting_with_special_tokens(self):
...
+
+        # NumPy scalar string.
+        output = tokenizer(np.str_("the quick"))
+        self.assertAllEqual(output, [1, 2, 3])
+
+        # NumPy scalar bytes.
+        output = tokenizer(np.bytes_("the quick"))
+        self.assertAllEqual(output, [1, 2, 3])
+
+    def test_python_workflow_with_seq_len_returns_dense_output(self):
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer_test.py:442]: missing WordPieceTokenizerTFTest subclass re-running the suite with _allow_python_workflow=False

File: keras_hub/src/tokenizers/word_piece_tokenizer_test.py [hunk 5]
```diff
@@ -250,6 +251,200 @@ def test_no_splitting_with_special_tokens(self):
...
+            _allow_python_workflow=False,
+        )
+
+        output = tokenizer("the quick")
+        self.assertAllEqual(output, [1, 2, 3])
+
+        output = tokenizer.detokenize([1, 2, 3])
+        self.assertAllEqual(output, "the quick")
+
+    def test_python_and_tf_workflows_match(self):
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer_test.py:308]: parity test too thin (add CJK, accents, special tokens, unicode whitespace, >100-byte words)

File: keras_hub/src/tokenizers/word_piece_tokenizer.py [hunk 6]
```diff
@@ -408,14 +592,28 @@ def set_vocabulary(self, vocabulary):
...
-            unknown_token=self.oov_token,
-            no_pretokenization=True,
-            support_detokenization=True,
-        )
+        # Build the python-side vocabulary lookups unconditionally. These
+        # are cheap to build and are used by the python tokenization
+        # workflow (`_tokenize_python`/`_detokenize_python`), which does not
+        # require `tensorflow_text` to be installed.
+        self._vocabulary_set = set(self.vocabulary)
+        self._token_to_id_python = {
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer.py]: duplicate-token token_to_id semantics

File: keras_hub/src/tokenizers/word_piece_tokenizer.py [hunk 7]
```diff
@@ -516,15 +722,229 @@ def tokenize(self, inputs):
...
+                + [pad_value] * (self.sequence_length - len(tokens))
+                for tokens in batched_tokens
+            ]
+
+        if unbatched:
+            tokens = batched_tokens[0]
+        else:
+            tokens = batched_tokens
+        if output_is_int and self.sequence_length:
+            return np.asarray(tokens, dtype=self.compute_dtype)
```
Comment (divyashreepathihalli) [inline keras_hub/src/tokenizers/word_piece_tokenizer.py]: np.asarray → backend tensor via __call__

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

File: keras_hub/src/tokenizers/word_piece_tokenizer.py (hunk 1 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer.py:311]: Hi @divyashreepathihalli , Thank you for your Review!

Separated whitespace patterns into `WHITESPACE_REGEX_TF`(using standard ASCII characters for RE2/TensorFlow Text compatibility) and 
`WHITESPACE_REGEX_PYTHON`(using the regex module with the scoped ASCII flag (?a:\s)|\p{Cc}|\p{Cf}) to ensure exact parity across both workflows without breaking RE2 compilation.

File: keras_hub/src/tokenizers/word_piece_tokenizer.py (hunk 2 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer.py]: Updated the Python case-folding logic to apply `NFKC normalization` via `unicodedata.normalize("NFKC", word.casefold())`, matching the behavior of `tf_text.case_fold_utf8`.

File: keras_hub/src/tokenizers/word_piece_tokenizer.py (hunk 3 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer.py]: Replaced character length bounds with UTF-8 byte-length checks `(len(word.encode("utf-8")) > 100)` in the pure Python WordPiece loop.

File: keras_hub/src/tokenizers/word_piece_tokenizer_test.py (hunk 4 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer_test.py:442]: Added the `WordPieceTokenizerTFTest` test case subclass to the test suite to ensure the entire test set runs and passes with `_allow_python_workflow=False`.

File: keras_hub/src/tokenizers/word_piece_tokenizer_test.py (hunk 5 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer_test.py:308]: Expanded `test_python_and_tf_workflows_match` to validate parity across `CJK` characters, `accent stripping`, `special tokens`, `unicode whitespace(\u00a0),` and >100-byte words.

File: keras_hub/src/tokenizers/word_piece_tokenizer.py (hunk 6 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer.py]: Fixed.

File: keras_hub/src/tokenizers/word_piece_tokenizer.py (hunk 7 above)
Comment (maitry63) [reply keras_hub/src/tokenizers/word_piece_tokenizer.py]: Replaced raw NumPy conversion `(np.asarray)` with Keras tensor `(keras.ops.convert_to_tensor)`.

**Comment by laxmareddyp** [comment]
closing this in favor of #3066
